### ISSUE TITLE
[OpenGL] NVIDIA GLSL stucks when compile or link some shaders

### ISSUE THREAD

**Issue by archibate** [taichi-dev/taichi#804]
TODO(archibate): Fix this after #785 merged.

**Describe the bug**
When running `examples/euler.py`, stuck at `glLinkProgram` when compiling:
```glsl
void calc_dt_c14_00() {
  float B = 100000.0;
  int D = 0;
  int AAk = 0;
  int F = D + 88080388 * AAk;
  int G = F + 16777216;
  int I = G + 4 * AAk;
  int J = I + 0;
  _data_f32_[J >> 2] = B;
}
```
If I move `dt` to the first, then `16777216` become `0` and compile passes to another problem.

**Log/Screenshots**
```
(dyn) [bate@archit taichi]$ TI_LOG_LEVEL=trace TI_ARCH=opengl op ../gpu/euler.py       
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-qjmdva_9
[Taichi] sandbox prepared
[T 04/17/20 09:50:17.976] [cuda_driver.cpp:CUDADriver@44] CUDA driver API (v10.2) loaded.
[Taichi] <dev mode>, supported archs: [cpu, cuda, opengl], commit 5f136cab, python 3.8.2
[T 04/17/20 09:50:18.253] [../gpu/euler.py:<module>@6] Unified memory disabled (env TI_USE_UNIFIED_MEMORY=0). This is experimental.                                                       
Following TI_ARCH setting up for arch=opengl
[T 04/17/20 09:50:18.253] [memory_pool.cpp:MemoryPool@9] Memory pool created. Default buffer size per allocator = 1024 MB                                                                 
[T 04/17/20 09:50:18.253] [llvm_context.cpp:TaichiLLVMContext@45] Creating Taichi llvm context for arch: x64                                                                              
[T 04/17/20 09:50:18.253] [llvm_context.cpp:TaichiLLVMContext@69] Taichi llvm context created.                                                                                            
[T 04/17/20 09:50:18.253] [program.cpp:Program@125] Program arch=opengl
[T 04/17/20 09:50:18.267] [/home/bate/Develop/taichi/python/taichi/lang/kernel.py:materialize@220] Materializing layout...                                                                
[T 04/17/20 09:50:18.280] [llvm_context.cpp:compile_runtime_bitcode@135] Compiling runtime module bitcode...                                                                              
[T 04/17/20 09:50:18.882] [llvm_context.cpp:compile_runtime_bitcode@150] runtime module bitcode compiled.                                                                                 
[T 04/17/20 09:50:19.060] [program.cpp:materialize_layout@265] materialize_layout called
[I 04/17/20 09:50:19.060] [program.cpp:materialize_layout@290] OpenGL root buffer size: 88080388 B                                                                                        
[W 04/17/20 09:50:19.060] [opengl_api.cpp:initialize_opengl@250] OpenGL backend currently WIP, MAY NOT WORK
[I 04/17/20 09:50:19.161] [opengl_api.cpp:initialize_opengl@276] [glsl] OpenGL 4.3.0 NVIDIA 440.64                                                                                        
[I 04/17/20 09:50:19.161] [opengl_api.cpp:initialize_opengl@278] [glsl] GLSL 4.30 NVIDIA via Cg compiler                                                                                  
[I 04/17/20 09:50:19.161] [opengl_extension.inc.h:initialize_opengl@3] [glsl] Found GL_ARB_compute_shader                                                                                 
[I 04/17/20 09:50:19.161] [opengl_extension.inc.h:initialize_opengl@4] [glsl] Found GL_ARB_gpu_shader_int64                                                                               
[I 04/17/20 09:50:19.161] [opengl_extension.inc.h:initialize_opengl@5] [glsl] Found GL_NV_shader_atomic_float                                                                             
[I 04/17/20 09:50:19.161] [opengl_extension.inc.h:initialize_opengl@7] [glsl] Found GL_NV_shader_atomic_int64                                                                             
[T 04/17/20 09:50:19.297] [/home/bate/Develop/taichi/python/taichi/lang/kernel.py:__call__@414] Compiling kernel set_ic_c8_0...                                                           
[D 04/17/20 09:50:19.320] [opengl_api.cpp:display_kernel_info@302] source of kernel [set_ic_c8_00] * 683:                                                                                 
......
void main()                                                                                  
{                                                                                            
  set_bc_c10_00();                                                                           
}                                                                                            
layout(local_size_x = 1536, local_size_y = 1, local_size_z = 1) in;                          

[T 04/17/20 09:50:19.350] [opengl_api.cpp:compile@87] glCompileShader IN
[T 04/17/20 09:50:19.350] [opengl_api.cpp:compile@89] glCompileShader OUT
[T 04/17/20 09:50:19.350] [opengl_api.cpp:link@127] glLinkProgram IN
[T 04/17/20 09:50:19.350] [opengl_api.cpp:link@129] glLinkProgram OUT
[T 04/17/20 09:50:19.351] [/home/bate/Develop/taichi/python/taichi/lang/kernel.py:__call__@414] Compiling kernel calc_dt_c14_0...                                                         
[D 04/17/20 09:50:19.360] [opengl_api.cpp:display_kernel_info@302] source of kernel [calc_dt_c14_00] * 1:                                                                                 
#version 430 core                                                                            
#extension GL_ARB_compute_shader: enable                                                     
#extension GL_ARB_gpu_shader_int64: enable                                                   
#extension GL_NV_shader_atomic_float: enable                                                 
#extension GL_NV_shader_atomic_int64: enable                                                 
precision highp float;                                                                       
layout(packed, binding = 0) buffer data_i32 { int _states_[2]; int _data_i32_[]; };          
layout(packed, binding = 0) buffer data_f32 { int _unused1_[2]; float _data_f32_[]; };       
layout(packed, binding = 0) buffer data_f64 { int _unused2_[2]; double _data_f64_[]; };      

void calc_dt_c14_00() {
  float B = 100000.0;
  int D = 0;
  int AAk = 0;
  int F = D + 88080388 * AAk;
  int G = F + 16777216;
  int I = G + 4 * AAk;
  int J = I + 0;
  _data_f32_[J >> 2] = B;
}

void main()
{
  calc_dt_c14_00();
}
layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;

[T 04/17/20 09:50:19.360] [opengl_api.cpp:compile@87] glCompileShader IN
[T 04/17/20 09:50:19.377] [opengl_api.cpp:compile@89] glCompileShader OUT
[T 04/17/20 09:50:19.377] [opengl_api.cpp:link@127] glLinkProgram IN
^C[ 2873.483099] [WARN]Received Interrupt signal.
^C[ 2874.114826] [WARN]Received Interrupt signal.
^C[ 2874.370854] [WARN]Received Interrupt signal.
^C[ 2874.602908] [WARN]Received Interrupt signal.
^C[ 2874.810948] [WARN]Received Interrupt signal.
^Z
[1]+  已停止               TI_LOG_LEVEL=trace TI_ARCH=opengl optirun python ../gpu/euler.py
```

**To Reproduce**
Run euler.py in #796 with NVIDIA GL.

**If you have local commits (e.g. compile fixes before you reproduce the bug), please make sure you first make a PR to fix the build errors and then report the bug.**

**Comment by archibate** [comment]
Seems moving the small tensors to the first fix the problem? Also in mpm128, it's the attractor_pos whose shape=() hurts NV GL.

**Comment by archibate** [comment]
Also found work well with `N = 128`.

**Comment by archibate** [comment]
Solution 1: sort snodes, from small to large.

**Comment by archibate** [comment]
Fixed in #808.
